rasterio/_vsiopener.pyx

```diff
@@ -77,6 +81,114 @@ cdef void uninstall_pyopener_plugin(VSIFilesystemPluginCallbacksStruct *callback
     callbacks_struct = NULL
 
 
+cdef int pyopener_stat(void *pUserData, const char *pszFilename, VSIStatBufL *pStatBuf, int nFlags) except -1 with gil:
+    if pUserData is NULL:
+        CPLError(CE_Failure, <CPLErrorNum>1, <const char *>"%s", <const char *>"Python opener is not initialized.")
+        return -1
+
+    cdef object var = <object>pUserData
+    cdef dict registry = var.get()
+    urlpath = pszFilename.decode("utf-8")
+    # The code below is just a test that we can assemble an st_mode
+    # value for GDAL from Python filesystem is_file and is_dir methods.
+    # Eventually we're going to dispatch to methods of the opener that
+    # was registered for the filename.
+    import stat
+    #try:
+    from urllib.parse import urlparse
+    parsed_uri = urlparse(urlpath)
+    path_to_check = Path(parsed_uri.path)
+    parent = path_to_check.parent
+    # Note that "r" mode is used here under the assumption that GDAL
+    # doesn't read_dir when writing data. Could be wrong!
+    mode = "rb"
+    key = ((parsed_uri.scheme, parsed_uri.netloc, parent.as_posix()), mode[0])
+    log.debug("Looking up opener in stat: registry=%r, urlpath=%r, mode=%r", registry, urlpath, mode)
+
+    try:
+        file_opener = registry[key]
+    except KeyError as err:
+        log.debug("Opener not found in registry: registry=%r, urlpath=%r, mode=%r", registry, urlpath, mode)
+        errmsg = f"Opener failed to open file with arguments ({repr(urlpath)}, {repr(mode)}): {repr(err)}"
+        errmsg_b = errmsg.encode("utf-8")
+        # 4 is CPLE_OpenFailedError.
+        CPLError(CE_Failure, <CPLErrorNum>4, <const char *>"%s", <const char *>errmsg_b)
+        return -1
+
+    try:
+        size = file_opener.size(urlpath)
+        if file_opener.is_file(urlpath):
+            fmode = 0o170000 | stat.S_IFREG
+        elif file_opener.is_dir(urlpath):
+            fmode = 0o170000 | stat.S_IFDIR
+        else:
+            fmode = 0
+    except (FileNotFoundError, KeyError):
+        # No such file or directory.
+        return 1
+    except Exception as err:
+        errmsg = f"Opener failed to determine file info: {repr(err)}"
+        errmsg_b = errmsg.encode("utf-8")
+        # 4 is CPLE_OpenFailedError.
+        CPLError(CE_Failure, <CPLErrorNum>4, <const char *>"%s", <const char *>errmsg_b)
+        return -1
+    else:
+        pStatBuf.st_size = size
+        pStatBuf.st_mode = fmode
```

Comment:
FYI, in very rare circumstances, st_mtime (modification time) can be used. Actually the only significant case where this is used is on the vector side, for the GML driver, where the GML driver checks the mtime of the .gml file vs the mtime of the .gfs file it generates the first time it opens the file.

Reply:
@rouault thanks for the heads up, I appreciate it. I wish GDAL wouldn't write files like .gfs and .aux.xml on a read. It would be simpler to work with GDAL if reads didn't have side effects.